flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java

```diff
@@ -171,6 +168,10 @@ private int selectNextReadingInputIndex() throws IOException {
 		updateAvailability();
 		checkInputSelectionAgainstIsFinished();
 
+		if (inputSelectionHandler.isInputUnavailable(0) && inputSelectionHandler.isInputUnavailable(1)) {
+			fullCheckAndSetAvailable();
+		}
```

Comment:
Again as explained above, `StreamTwoInputProcessor#updateAvailability()` should do the trick (% the starvation check in L181)

Reply:
The commit is unnecessary. I did not fully understand the availability handling.